c++/defs/repositories.bzl

```diff
@@ -0,0 +1,78 @@
+"Provides dependency definition macros for Cap'n'Proto, and for use in downstream Bazel projects."
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+load(
+    "//defs:dependencies.bzl",
+    "BROTLI",
+    "ZLIB",
+    "SKYLIB",
+    "BORINGSSL",
+)
+load(
+    "@bazel_tools//tools/build_defs/repo:utils.bzl",
+    "maybe",
+)
+
+
+def capnp_repositories(
+    enable_bzlmod = False,
+    omit_brotli = False,
+    omit_skylib = False,
+    omit_zlib = False,
+    omit_boringssl = False):
+    """Add Cap'n'Proto-required dependencies for a Bazel project workspace
+
+    If repositories are already available in the Bazel build, they will not be re-added.
+    Pass `omit_*` to avoid a particular dependency.
+
+    Args:
+        enable_bzlmod: Only add dependencies not available via Bzlmod.
+        omit_brotli: Don't add a Brotli dependency.
+        omit_skylib: Don't add a Skylib dependency.
+        omit_zlib: Don't add a Zlib dependency.
+        omit_boringssl: Don't add a Boringssl dependency.
+    """
+
+    # Brotli is not available yet on the Bazel Registry.
+    if not omit_brotli:
+        maybe(
+            http_archive,
+            name = "brotli",
+            sha256 = BROTLI.digest,
+            strip_prefix = "brotli-%s" % BROTLI.version,
+            type = "tgz",
+            urls = ["https://github.com/google/brotli/archive/refs/tags/v%s.tar.gz" % BROTLI.version],
+        )
```

Comment:
This is the `capnp_repositories` macro, which downstream users use in their `WORKSPACE` file. **This needs to be loaded and called for every downstream project,** including projects which use Bzlmod, because Brotli isn't available on BCR yet.

If users have some of these installed already, or would encounter conflicts if Cap'n'Proto installed them too early in their `WORKSPACE`, they can pass `omit_*` to prevent a dependency from being added (for example, `omit_brotli = True`). Dependencies are added with `maybe(...)` so they won't overwrite existing ones anyway.

For Bzlmod users who must call this function, `enable_bzlmod = True` is passed, which _omits all dependencies available on Bzlmod_, so only Brotli is installed in that case. (looking at you, Brotli).

Usage example:

**In a `WORKSPACE` file**:
```starlark
load("@capnp-cpp//defs:workspace.bzl", "capnp_repositories")

capnp_repositories()
```

**For Bzlmod users, in a `WORKSPACE.bzlmod` file**:
```starlark
load("@capnp-cpp//defs:workspace.bzl", "capnp_repositories")

capnp_repositories(enable_bzlmod = True)
```

Reply:
Update: Nevermind? I guess Brotli is only used outside the main library, maybe? Because the integration tests didn't complain and didn't have this declaration. Open to advice from somebody who knows better how Brotli is used, maybe we can avoid this.

Reply:
Seemed to work fine for me in bzlmod 6.4 without this! 🤷‍♀️